app/upgrades/v4/msg_restrict_ubedrock_ante.go

```diff
@@ -0,0 +1,43 @@
+package v4
+
+import (
+	pylonskeeper "github.com/Pylons-tech/pylons/x/pylons/keeper"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+type MsgRestrictUbedrockDecorator struct {
+	pk pylonskeeper.Keeper
+}
+
+func NewMsgRestrictUbedrockDecorator(pk pylonskeeper.Keeper) MsgRestrictUbedrockDecorator {
+	return MsgRestrictUbedrockDecorator{
+		pk: pk,
+	}
+}
+
+// AnteDecorator for restrict ubedrock denom used by unallowed address
+func (ad MsgRestrictUbedrockDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if (ctx.IsCheckTx() || ctx.IsReCheckTx()) && !simulate {
+		sigTx, ok := tx.(authsigning.SigVerifiableTx)
+		if !ok {
+			return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "invalid transaction type")
+		}
+
+		messages := sigTx.GetMsgs()
+		if len(messages) <= 0 {
+			return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "invalid messages")
+		}
+
+		if msgSend, ok := messages[0].(*banktypes.MsgSend); ok {
+			if ok, _ = msgSend.Amount.Find("ubedrock"); ok {
+				if _, kycAcc_found := ad.pk.GetPylonsKYC(ctx, msgSend.ToAddress); kycAcc_found == false {
+					return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "'ubedrock' should only be transfer among allowed address")
```

Comment:
should only be "transferred", small typo

Reply:
Fixed in https://github.com/Pylons-tech/pylons/pull/1086/commits/9563984c5a81c22c0321c6f7d63962c9d7026564